test/interceptor_test.go

```diff
@@ -277,3 +279,409 @@ func (s) TestChainStreamClientInterceptor_ContextValuePropagation(t *testing.T)
+		if ctx.Value(secondInterceptorCtxkey{}) != nil {
+			errCh.SendContext(ctx, fmt.Errorf("first interceptor should not have %T in context", secondInterceptorCtxkey{}))
+		}
+		firstCtx := context.WithValue(ctx, firstInterceptorCtxkey{}, firstInterceptorCtxVal)
+		return invoker(firstCtx, method, req, reply, cc, opts...)
+	}
+
+	secondInt := func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		if got, ok := ctx.Value(parentCtxkey{}).(string); !ok || got != parentCtxVal {
+			errCh.SendContext(ctx, fmt.Errorf("second interceptor got %q in context.Val, want %q", got, parentCtxVal))
+		}
+		if got, ok := ctx.Value(firstInterceptorCtxkey{}).(string); !ok || got != firstInterceptorCtxVal {
+			errCh.SendContext(ctx, fmt.Errorf("second interceptor got %q in context.Val, want %q", got, firstInterceptorCtxVal))
+		}
+		if ctx.Value(secondInterceptorCtxkey{}) != nil {
+			errCh.SendContext(ctx, fmt.Errorf("second interceptor should not have %T in context", secondInterceptorCtxkey{}))
+		}
+		secondCtx := context.WithValue(ctx, secondInterceptorCtxkey{}, secondInterceptorCtxVal)
+		return invoker(secondCtx, method, req, reply, cc, opts...)
+	}
+
+	lastInt := func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		if got, ok := ctx.Value(parentCtxkey{}).(string); !ok || got != parentCtxVal {
+			errCh.SendContext(ctx, fmt.Errorf("last interceptor got %q in context.Val, want %q", got, parentCtxVal))
+		}
+		if got, ok := ctx.Value(firstInterceptorCtxkey{}).(string); !ok || got != firstInterceptorCtxVal {
+			errCh.SendContext(ctx, fmt.Errorf("last interceptor got %q in context.Val, want %q", got, firstInterceptorCtxVal))
+		}
+		if got, ok := ctx.Value(secondInterceptorCtxkey{}).(string); !ok || got != secondInterceptorCtxVal {
+			errCh.SendContext(ctx, fmt.Errorf("last interceptor got %q in context.Val, want %q", got, secondInterceptorCtxVal))
+		}
+		errCh.SendContext(ctx, nil)
+		return invoker(ctx, method, req, reply, cc, opts...)
+	}
+
+	// Start a stub server.
+	ss := &stubserver.StubServer{
+		EmptyCallF: func(context.Context, *testpb.Empty) (*testpb.Empty, error) { return &testpb.Empty{}, nil },
+	}
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Failed to start stub server: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	opts := []grpc.CallOption{
+		grpc.PerRPCUnaryClientInterceptor(firstInt),
+		grpc.PerRPCUnaryClientInterceptor(secondInt),
+		grpc.PerRPCUnaryClientInterceptor(lastInt),
+	}
+	// Use the above chain of interceptors while invoking a unary RPC.
+	if _, err := ss.Client.EmptyCall(context.WithValue(ctx, parentCtxkey{}, parentCtxVal), &testpb.Empty{}, opts...); err != nil {
+		t.Fatalf("ss.Client.EmptyCall() failed: %v", err)
+	}
+	val, err := errCh.Receive(ctx)
```

Comment:
I think we can safely ignore the returned value from this `Receive` since we are only concerned about whether we receive something or not. So, this can be replaced with something like:

```golang
	_, err := errCh.Receive(ctx); if err != nil {
		t.Fatalf("timeout when waiting for unary interceptor to be invoked: %v", err)
	}
```

And skip the check for `if val != nil {...}`.

Same applies for other tests as well.

Reply:
Yeah, I've been thinking about all tests in `interceptor_test.go` file. There are two things that bother me:
1) They don't actually check if an interceptor is called. They check only if the last one is called so they assume that when we reach the last interceptor then all predecessors are called. I'm not sure if this is a good assumption. It would be better to not rely on it and check if each of them is called as expected.
2) If one of the checks fails then we send an error to `errCh`. This is a buffered with size 1 so then when we try to write to this channel again (in the last interceptor to acknowledge the end of the chain) we're blocked because reading from `errCh` happens after receiving an RPC so `ctx` times out and we always get a `DeadlineExceeded` error from the RPC, not the real one sent to `errCh`.

Do you think those deserve to be fixed?

Reply:
I think it would nice to have these fixed. I'm guessing if we have a separate channel for each of those interceptors, we can fix both of these issues. And we can fix these as a separate PR. Thank you.